Improve: add functions for handling game character details

SlySven opened the pull request:
~~`getCharacterName` takes no arguments and returns the "Character name" from the connection screen. There is no provision to change this stored setting.~~ *Extracted and implemented separately as PR #6023.*

`sendCharacterName` takes no arguments and causes the "Character name" from the connection screen to be sent without echoing to the game server, it will return a nil plus an error message should the name not be set (is empty string).

`sendCharacterPassword` takes no arguments and causes the "Password" from the connection screen to be sent without echoing to the game server, it will return a nil plus an error message should no password be set (is empty string). There is no provision to find out this stored setting.

`sendCustomLoginText` takes no arguments and causes the custom login text to be sent provided it is not disabled (set to the 0th one).

`getCustomLoginTextId` gets the index (from zero upwards) of the custom login text - so at this point will return `0` {disabled} or `1`.

Remove saving and restoring login name and password either side of package/module loading - we do not need them to protect the details from being changed in a loaded XML file because those pieces of information are not stored in there.

Move the existing timed sending of the user name and password to a new `doLogin()` Lua function in the external `LuaGlobal.lua` file which effectively does the same thing but can be replaced by something more sophisticated if the user needs it. It also checks whether the option to use a custom login text has been set and will send the chosen string two seconds after a successful connection instead of the character name and password after two and three seconds respectively.

Alternatively the user can replace it with a dummy (empty) function with the same name and use a pair of triggers (which I have tested) or an `sysConnectionEvent` event handler.

The custom login currently only offers one text - of the form:
"connect {character name} {password}"
which can be selected from a new comboBox on the "Special options" tab of the preferences. It is saved with the profile data.

The custom strings could also be appended to if a compelling case is made to the Mudlet Makers though for backwards compatibility it would not be recommended to remove a string previously added.

For user display purposes the string shown in the preferences is expected to be translated for the tokens "{character name}" and "{password}" however the system is designed to not have the actual string changed on a per locale basis. The UI texts are generated by a method that detects whether it is being used for the first time on dialogue creation (and so considers the existing profile's setting) or after a GUI language change - which keeps the same setting but shows it in the new language. It is my intention to refactor out the population of some of the other combo-boxes in that dialogue as well so that the user will not have to restart Mudlet just to get other important settings in a language that they can comprehend on the first time they run Mudlet.

Also:
* demote dlgConnectionProfiles::slot_connectToServer() from a slot to a normal method - as it is not connected up to the signal/slot system.

This is a cherry-pick and edit of code originally promulgated as #2252 with the extraneous bits picked out...! :grinning: 

Closes #2961, #3920 and would also have closed #3562 if it were still open.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: add functions for handling game character details

`getPlayerName` takes no arguments and returns the "Character name" from
the connection screen. There is no provision to change this stored setting.

`sendPlayerName` takes no arguments and causes the "Character name" from
the connection screen to be sent without echoing to the game server, it
will return a nil plus an error message should the name not be set (is
empty string).

`sendPlayerPassword` takes no arguments and causes the "Password" from
the connection screen to be sent without echoing to the game server, it
will return a nil plus an error message should no password be set (is empty
string). There is no provision to find out this stored setting.

`sendCustomLoginText` takes no arguments and causes the custom login text
to be sent provided it is not disabled (set to the 0th one).

`getCustomLoginTextId` gets the index (from zero upwards) of the custom
login text - so at this point will return `0` {disabled} or `1`.

Remove saving and restoring login name and password either side of
package/module loading - we do not need them to protect the details from
being changed in a loaded XML file because those pieces of information are
not stored in there.

Move the existing timed sending of the user name and password to a new
`doLogin()` Lua function in the external `LuaGlobal.lua` file which
effectively does the same thing but can be replaced by something more
sophisticated if the user needs it. It also checks whether the option to
use a custom login text has been set and will send the chosen string two
seconds after a successful connection instead of the character name and
password after two and three seconds respectively.

Alternatively the user can replace it with a dummy (empty) function with
the same name and use a pair of triggers (which I have tested) or an
`sysConnectionEvent` event handler.

The custom login currently only offers one text - of the form:
"connect {character name} {password}"
which can be selected from a new comboBox on the "Special options" tab of
the preferences. It is saved with the profile data.

The custom strings could also be appended to if a compelling case is made
to the Mudlet Makers though for backwards compatibility it would not be
recommended to remove a string previously added.

For user display purposes the string shown in the preferences is expected
to be translated for the tokens "{character name}" and "{password}" however
the system is designed to not have the actual string changed on a per
locale basis. The UI texts are generated by a method that detects whether
it is being used for the first time on dialogue creation (and so considers
the existing profile's setting) or after a GUI language change - which
keeps the same setting but shows it in the new language. It is my intend
to refactor out the population of some of the other combo-boxes in that
dialogue as well so that the user will not have to restart Mudlet just to
get other important settings in a language that they can comprehend.

Also:
* demote dlgConnectionProfiles::slot_connectToServer() from a slot to a
normal method - as it is not connected up to the signal/slot system.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
> It is my intention to refactor out the population of some of the other combo-boxes in that dialogue as well so that the user will not have to restart Mudlet just to get other important settings in a language that they can comprehend on the first time they run Mudlet.

That is not a practical problem facing Mudlet players today - let's avoid complicating the code further to address a non-issue. There are other real issues [we could fix](https://github.com/Mudlet/Mudlet/issues?q=is%3Aopen+is%3Aissue+label%3Abug) instead.

vadi2 commented on src/Host.cpp:
```diff
@@ -46,6 +46,8 @@
 #include <memory>
 #include "post_guard.h"
 
+const int Host::mLuaSendPasswordTimeout = 30000; // milli-seconds!
```
```suggestion
using namespace std::chrono_literals;

const int Host::mLuaSendPasswordTimeout = 30s;
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -18304,6 +18309,95 @@ int TLuaInterpreter::getDictionaryWordList(lua_State* L)
+    const QString name{host.getLogin()};
+    if (name.isEmpty()) {
+        lua_pushnil(L);
+        lua_pushstring(L, "no character name set, so nothing to return");
```
```suggestion
        lua_pushstring(L, "no character name set");
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -18304,6 +18309,95 @@ int TLuaInterpreter::getDictionaryWordList(lua_State* L)
+
+    if (host.getLogin().isEmpty()) {
+        lua_pushnil(L);
+        lua_pushstring(L, "no character name set, so nothing to send");
```
```suggestion
        lua_pushstring(L, "no character name set");
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -18304,6 +18309,95 @@ int TLuaInterpreter::getDictionaryWordList(lua_State* L)
+    // something but was just out of stock in the password department! 8-)
+    if (host.getPass().isEmpty()) {
+        lua_pushnil(L);
+        lua_pushstring(L, "no password set, so nothing to send");
```
```suggestion
        lua_pushstring(L, "no password set");
```

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -149,13 +151,28 @@ cTelnet::cTelnet(Host* pH, const QString& profileName)
+    connect(this, &cTelnet::signal_disconnected, this, &cTelnet::slot_disableLuaSendPassword);
+    connect(mpLuaSendPasswordTimer, &QTimer::timeout, this, &cTelnet::slot_disableLuaSendPassword);
+
+#if defined (QT_DEBUG)
```
:+1:

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -440,44 +463,99 @@ void cTelnet::handle_socket_signal_error()
+    if (loginString.contains(QLatin1String("{character name}"), Qt::CaseSensitive)) {
+        const QString characterName = mpHost->getLogin();
+        if (characterName.isEmpty()) {
+            return {false, QStringLiteral("custom login %1 requires the character name and that is not currently defined for this profilee").arg(actualLoginId)};
```
`profilee` typo here and below.

SlySven replied:
:man_facepalming:

vadi2 reviewed (commented):
This a great start for custom login texts, but we need to streamline the design to be simpler to understand for everyone involved. 

I don't understand how can I make a custom login setup for a game that doesn't use "connect username password" when the options in the UI fit neither.

I had a much simpler design in my head: no UI changes, just a `doLogin()` function that you can overwrite. People then can create a package that'll handle logging in appropriately. For example we could make a "connect username password" package - install it and you're good to go. Combined with the fact that we're working on the package manager, having more extensibility and less things hardcoded in Mudlet is better this way.

SlySven commented:
> I don't understand how can I make a custom login setup for a game that doesn't use "connect username password" when the options in the UI fit neither.
>
> I had a much simpler design in my head: no UI changes, just a doLogin() function that you can overwrite. People then can create a package that'll handle logging in appropriately. For example we could make a "connect username password" package - install it and you're good to go. Combined with the fact that we're working on the package manager, having more extensibility and less things hardcoded in Mudlet is better this way.

The whole point is to avoid the possibility of a script/package being put in that could send a random (i.e. the contents not reviewed/approve by us) string containing the password - and for the use of that to be clearly specified/chosen/visible/controllable to/by the user via the GUI.

SlySven committed:
Update src/Host.cpp

Vadim pointed out how to use std::chrono_literals - which I did not previous understand...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

SlySven committed:
Revise: shorten some error messages

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

vadi2 commented:
We can solve that  just by disabling the send password function after X seconds within connecting.

SlySven commented on src/ctelnet.cpp:
```diff
@@ -440,44 +463,99 @@ void cTelnet::handle_socket_signal_error()
+    if (loginString.contains(QLatin1String("{character name}"), Qt::CaseSensitive)) {
+        const QString characterName = mpHost->getLogin();
+        if (characterName.isEmpty()) {
+            return {false, QStringLiteral("custom login %1 requires the character name and that is not currently defined for this profilee").arg(actualLoginId)};
```
```suggestion
            return {false, QStringLiteral("custom login %1 requires the character name and that is not currently defined for this profile").arg(actualLoginId)};
```

vadi2 commented on src/Host.cpp:
```diff
@@ -46,7 +46,9 @@
-const int Host::mLuaSendPasswordTimeout = 30000; // milli-seconds!
+using namespace std::chrono_literals;
+
+const int Host::mLuaSendPasswordTimeout = 30s;
```
Sorry, it shouldn't be an int, try auto.

SlySven replied:
And what goes in for the declaration in the header file?

SlySven commented on src/ctelnet.cpp:
```diff
@@ -440,44 +463,99 @@ void cTelnet::handle_socket_signal_error()
+            return {false, QStringLiteral("custom login %1 requires the character name and that is not currently defined for this profilee").arg(actualLoginId)};
+        }
+        if (password.isEmpty()) {
+            return {false, QStringLiteral("custom login %1 requires the password and that is not currently defined for this profilee").arg(actualLoginId)};
```
```suggestion
            return {false, QStringLiteral("custom login %1 requires the password and that is not currently defined for this profile").arg(actualLoginId)};
```

SlySven commented on src/ctelnet.cpp:
```diff
@@ -440,44 +463,99 @@ void cTelnet::handle_socket_signal_error()
+        loginString = loginString.arg(characterName, password);
+    } else {
+        if (password.isEmpty()) {
+            return {false, QStringLiteral("custom login %1 requires the password and that is not currently defined for this profilee").arg(actualLoginId)};
```
```suggestion
            return {false, QStringLiteral("custom login %1 requires the password and that is not currently defined for this profile").arg(actualLoginId)};
```

SlySven committed:
Revise: correct typoo in error message

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: fix use of std::chrono elements to specify timer interval

Using `auto` doesn't seem to work in this situation but it isn't the end
of the work to have to give the type in the definition.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven commented:
That last commit was needed to resolve a conflict caused by another (just merged by me) PR adding something else in the same place in a header file...

vadi2 commented:
https://github.com/Mudlet/Mudlet/pull/3952#pullrequestreview-442187029 is still outstanding :) I'd like to simplify the design to be clearer for everyone involved.

SlySven commented:
The design herein is such that the password is not made available - by design - to the Lua sub-system - so you cannot write a replacement `doLogin()` to include it.

Either you have a situation where you can send it on it's own as text (with an EOL tagged on obviously) which will be quite hard (though I would not claim for it to be impossible) for another player to see; or a preselected text with it (and potentially the character name) is sent. A lua script can check that the text is the one that a particular MUD might need but it cannot change it - only the user can do that and they have to consciously enable it.

I personally consider that this is a reasonably secure way of implementing this feature and there are some parties that have expressed an interest in the fixed string extension that has already made this PR's scope creep. I am not sure whether any further extension or revision is needed at this point - barring something for that party that was sniffing out the password via a hole that is to be blocked off in the next release so that they could send it somewhere not even in an Out-Of-Band Telnet protocol!

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

Conflicts resolved in:
* src/ui/profile_preferences.ui

This was caused by both arms of the merge both modifying this file and it
was necessary to tweak things so that the changes in this PR fitted in
with the already committed PR that sorted out the control order and
rearranged some controls and corrected the tab order...

Also added in missing tabstop for the recently added force CHARSET
negotiation off so it was tabbed to in the right order!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: reorganise so that this can merge with other changes

This PR will clash with #3832 so this commit tweaks the position (row) of
a couple of controls so that they will merge with another new control in
that PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

vadi2 commented:
I think it's OK to make it available within the first X seconds of connecting, and then the function doesn't do anything. That would solve it, wouldn't it?

SlySven commented:
> I think it's OK to make it available within the first X seconds of connecting, and then the function doesn't do anything. That would solve it, wouldn't it?

You mean have a `getCharacterPassword()` function (which only works for a short time) - well that could reveal the profile's password to any script that had hooked into the `sysLoadEvent` - that would blow apart the sandboxing of the password that I'm trying to maintain here - as once a mal-package (?!) has gotten hold of the password it could do anything with it at any time afterwards... 🙁

vadi2 commented:
Why - we don't have to pass what `sendCharacterPassword()` sends into the `sysDataSendRequest`, right?

vadi2 commented:
@tamarindo maybe your security knowledge could assist us here? 

We'd like to make it possible to have a custom login string (to support `connect charname password` for example) as simple as possible, while being secure so scripts can't intercept the password.

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
BugFix: remove trace of QMutex that is not longer used

This was left over when I did an on-line merge conflict removal as I was
respecting a mutex in my code that is not there anymore!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

mpconley commented:
This is beyond me at the moment from the text above.  Would a threat model help here?  https://github.blog/2020-09-02-how-we-threat-model/

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

Conflicts resolved in:
* src/XMLimport.cpp
* src/ctelnet.cpp
* src/ctelnet.h
* src/dlgConnectionProfiles.cpp
* src/dlgProfilePreferences.h

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:books: Wiki documentation updated to include `sendCustomLoginText()` and `getCustomLoginTextId()`.

vadi2 reviewed (commented):
Getting a crash when connecting with username/password filled in because `mTimerLogin` is `nullptr`, could you look?

![image](https://user-images.githubusercontent.com/110988/112189459-547a6f80-8c04-11eb-9829-ae6c04f54bde.png)

SlySven commented:
Those timers shouldn't exist any longer - I removed them - somehow a merge has gone wrong and put them back...

:roll_eyes:

SlySven committed:
FixUp: correct something that looks like a merge error

Somehow some lines of code that this PR removed had gotten reinserted
during a previous merge!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
Revise: keep GHA Clang-tidy happy

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/Host.h:
```diff
@@ -198,6 +198,8 @@ class Host : public QObject
                         useDictionary = mEnableUserDictionary;
                         useShared = mUseSharedDictionary; }
+    void            setCustomLoginId(const int value);
+    int             getCustomLoginId() { return mCustomLoginId; }
```
Done.

Kebap commented:
Moved preliminary documentation to https://wiki.mudlet.org/w/Area_51

SlySven committed:
Merge: update 'Enhance_addFunctionsForHandlingGameCharacterDetails_redux'

Pull in changes from 'development' branch and resolved conflicts in:
* src/ctelnet.cpp
* src/ctelnet.h
* src/dlgConnectionProfiles.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: reset to use normal alignment for control label

I.e. Left rather than right in most Styles.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: fix spelling and add issue cross-reference

Correct a spelling mistake "permenant" -> "permanant" in a comment.

Add details of a related Issue to cover a new TODO comment.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

vadi2 commented:
@SlySven what do you think about https://github.com/Mudlet/Mudlet/pull/3952#issuecomment-704491860 ?

SlySven commented:
>>> I think it's OK to make it available within the first X seconds of connecting, and then the function doesn't do anything. That would solve it, wouldn't it?
>>>
>>You mean have a `getCharacterPassword()` function (which only works for a short time) - well that could reveal the profile's password to any script that had hooked into the `sysLoadEvent` - that would blow apart the sandboxing of the password that I'm trying to maintain here - as once a mal-package (?!) has gotten hold of the password it could do anything with it at any time afterwards... :slightly_frowning_face:
>>
> @SlySven what do you think about #3952 (comment) ?

That would give the password to the Lua sub-system - and once a package or script has gotten hold of the password then it can anything it likes with it. The whole reason for the way I am have done things is to ensure *there is absolutely* **no way** *for that system to find out what the password is...*  What you are suggesting would blow a hole right out of that sand-pit!

vadi2 commented:
I don't see how. Did you read my comment about explicitly not passing it to Lua?

vadi2 commented:
Because we're in control of both the Lua subsystem and C++ core, we have the power of choosing what to show to Lua, solving this problem!

SlySven commented:
😕 I don't understand what you are saying now! I have always said that the Lua sub-system must not ever see the password - so that a function like `getPlayerPassword()` is not something I'd want to see in the Lua API - but that is what I thought you wanted - with the limitation that it would only work for a short while after the connection to the Server was made.

So, if we are in agreement on that - then what it is that is unsatisfactory about this proposal as it is now? To recap as part of this PR, the Lua API will allow the user to:
* find out what the player's character name is
* to send that name on it's own to the Game Server
* to send the password on it's own to the Game Server, but without ever knowing what it is - but only for a limited time after a connection is made
* to find out if a custom login text mode has been selected by the user (so a script package can check that the correct one has been chosen) at present only one such text has been inserted, any further ones would need to be put to us but I can't see that this will be requested anytime soon. {Perhaps with the exception of that one MUD that wants the details to be sent off to a special login server for their game but that will require something a bit more complicated including some sort of HTTPS message sending/receiving - that will be a bit more work and is more than just a simple text substitution but I think it would be doable}
* to request that the custom login text is sent including the password and optionally the username if that is part of the definition - again within the time limit after the connection has been made.

Importantly in all of this, the system will not allow a script to send out a random/ad hoc line of text which happens to contain the password - it can only be sent on its own or within a particular text that the user must select themselves i.e. it cannot be set by a script/package - though those can check that the wanted one has been chosen by the user.

SlySven commented:
As an example of this proposal in action, it already includes a default Lua login script that replaces/replicates the former C++:
```lua
--[[
Effectively this has migrated from cTelnet::handle_socket_signal_connected()
but can be replaced with something more advanced by the player if needed. It
would also be possible to produce a sysConnectionEvent handler or a simple
pair) of triggers. In either of those two cases it would be necessary to replace
this predefined function in the profile's scripts with a dummy, do nothing one,
of the form:

  function doLogin()
  end

so that the Lua interpreter doesn't run this code.

Also revised to check and use a custom login text if specified in the profile
preferences (on "Special options" tab):
]]
function doLogin()
  if getCustomLoginId() > 0 then
    tempTimer(2.0, [[sendCustomLoginText()]], 1)
  elseif getCharacterName() ~= "" then
    tempTimer(2.0, [[sendCharacterName()]], 1)
    tempTimer(3.0, [[sendCharacterPassword()]], 1)
  end
end
```

vadi2 commented:
This warrants a re-read of the whole PR 🕵️‍♂️

vadi2 commented:
OK, first of all, I'd like to apologise - I came off way strong in my first review!

I have an idea on how to build on what you've got to make it simpler and more inline with how Mudlet should be - let's have a call, it'll be easier to explain that way.

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

Needed to remove getCharacterName() function which has been extracted
from this PR and done separately - and now merged into the main
developement branch.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

This is so as to try and fix some CI issues that I believe have been
solved in the main development branch.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

To bring in changes from main development branch.

Signed-off by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

Kebap commented:
> let's have a call, it'll be easier to explain that way.

Did you have the call?

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

vadi2 committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

vadi2 commented:
We did not... and even with my simpler plan of implementing it, it would still have required the player to fobble with some knobs in order to get it to work.

That's not how we want Mudlet to be though, right? We want things to just work out of the box. So I put together a spec instead to address this: https://wiki.mudlet.org/w/Standards:GMCP_Authentication, and I implemented it in Mudlet: https://github.com/Mudlet/Mudlet/pull/7152. There's also [a little mock](https://github.com/vadi2/gmcp-auth-mock) server validating the implementation that can be used as inspiration for server developers.

If a server implements GMCP Auth, it will both avoid the players having to configure how authentication should work and it will avoid the security issues we've been running into as we try to implement this.

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
Fix: repair something that had broken off over life-time of PR

Also update some things that had been changed elsewhere in the code-base
during the period this PR has been open. (Replacement of `QStringLiteral`
by `qsl`, addition of `(QString) utils::richText(const QString)` wrapper.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Hey, appreciate the effort here. This might be more complex than we need at the moment 1, so I’m closing the PR. Let’s keep the ideas coming!

1 (in comparison, making a trigger is much simpler, although not as as secure as a password in a keychain)

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

Conflicts resolved in:
* src/dlgConnectionProfiles.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

SlySven committed:
Fix: adapt this PR to accomodate parallel changes made by #7152

Extended the functions disabled by the login timeout to include the
`sendCharacterName()` function - and allow the timer that provides that
timeout to be stopped by the GMCP authenticator that #7152 introduced. Such
a stop also immediately disables those functions so that it will neutralise
any scripts or triggers that would be using them.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_addFunctionsForHandlingGameCharacterDetails_redux

ZookaOnGit commented:
I use multiple characters on one profile so don't use autologin.  My password is shown in my logs and command line history.  It could also be sniffed with sysDataSendRequest.  It is sent plain text down the wire all the way to the game server.  Does my game encrypt it? Who knows?! There are other methods to get my password which could be used.  Spending a lot of effort here on security when there are alternatives to snatch it, just saying.

Just make it easier and customisable for the end user.  And for me this would mean triggers and scripts and `getCharacterName()` and `getCharacterPassword()`

SlySven commented:
@ZookaOnGit doing that is your choice :astonished: - but given the other steps we have taken over the years to store the passwords securely within each OSes "secure storage" systems if that is the (less portable, admittedly) choice they take and to provided means to use an encrypted transport to the Game Server if it is available then we don't want to make it trivial for them to be sniffed out by a "bad" package/script...

ZookaOnGit commented:
Seems discussion on the best way to move forward has stalled.  Closing this one for now.
